vm/native/token.go

```diff
@@ -0,0 +1,382 @@
+package native
+
+import (
+	"errors"
+	"fmt"
+	"github.com/iost-official/go-iost/core/contract"
+	"github.com/iost-official/go-iost/vm/host"
+	"math"
+	"strconv"
+)
+
+var tokenABIs map[string]*abi
+
+// const prefix
+const (
+	TokenInfoMapPrefix    = "TI"
+	TokenBalanceMapPrefix = "TB"
+	IssuerMapField        = "issuer"
+	SupplyMapField        = "supply"
+	TotalSupplyMapField   = "totalSupply"
+	CanTransferMapField   = "canTransfer"
+	DefaultRateMapField   = "defaultRate"
+	DecimalMapField       = "decimal"
+)
+
+func init() {
+	tokenABIs = make(map[string]*abi)
+	register(&tokenABIs, createTokenABI)
+	register(&tokenABIs, issueTokenABI)
+	register(&tokenABIs, transferTokenABI)
+	register(&tokenABIs, balanceOfTokenABI)
+	register(&tokenABIs, supplyTokenABI)
+	register(&tokenABIs, totalSupplyTokenABI)
+	register(&tokenABIs, destroyTokenABI)
+}
+
+func checkTokenExists(h *host.Host, tokenName string) (ok bool, cost *contract.Cost) {
+	exists, cost0 := h.MapHas(TokenInfoMapPrefix+tokenName, IssuerMapField)
+	return exists, cost0
+}
+
+func getBalance(h *host.Host, tokenName string, from string) (balance int64, cost *contract.Cost) {
+	balance = int64(0)
+	cost = contract.Cost0()
```

Comment:
in `contract.Cost0()`,  `Cost0`, the `0` seems like index, rename `Cost0` to `CostZero` or `CostEmpty` will be better, i think 🤔